src/txkube/_authentication.py

```diff
@@ -202,7 +203,7 @@ def creatorForNetloc(self, hostname, port):
 
         :see: ``twisted.web.iweb.IPolicyForHTTPS``
         """
-        hostname = hostname.decode("ascii")
+        hostname = ensure_unicode(hostname)
```

Comment:
This still looks mistaken to me.  See https://twistedmatrix.com/documents/current/api/twisted.web.iweb.IPolicyForHTTPS.html.  hostname is documented as bytes.

Reply:
This is needed.  Without this, I get:

```
Failure: testtools.testresult.real._StringException: Traceback (most recent call last):
  File "/Users/crodrigues/txkube/src/txkube/test/test_authentication.py", line 354, in test_creatorForNetLoc_interface
    def test_creatorForNetLoc_interface(self, host_known, host_used, port_known, port_used):
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/core.py", line 726, in wrapped_test
    state.run()
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/core.py", line 619, in run
    print_example=True, is_final=True
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/executors.py", line 78, in <lambda>
    lambda: function(data)
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/executors.py", line 33, in execute
    return function()
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/executors.py", line 78, in <lambda>
    lambda: function(data)
  File "/Users/crodrigues/venv-3.6/lib/python3.6/site-packages/hypothesis/core.py", line 115, in run
    return test(*args, **kwargs)
  File "/Users/crodrigues/txkube/src/txkube/test/test_authentication.py", line 368, in test_creatorForNetLoc_interface
    creator = policy.creatorForNetloc(host_used, port_used)
  File "/Users/crodrigues/txkube/src/txkube/_authentication.py", line 205, in creatorForNetloc
    hostname = hostname.decode("ascii")
AttributeError: 'str' object has no attribute 'decode'


```

Note, I discovered this in https://github.com/LeastAuthority/txkube/pull/162 where I fixed more stuff and got more of the tests to pass on Python 3.

Reply:
I agree with exarkun . this should be bytes... and not Unicode... the error is in another part

```
    hostname = hostname.decode("ascii")
AttributeError: 'str' object has no attribute 'decode'
```

this errors said: "hosname" is already Unicode, what more do you want from it? :)

A valid fix should make sure that you don't pass the hosname as Unicode.

These type of on-demand inner layer data conversion will not help us fix the str vs Unicode mess that we have in Python 2.

Is good that we see these type of errors as  they help identify the code which might have different behaviour between Python 2 and 3.

In this case, I think that we need to fix the test.
A fix might look like this:
```
        creator = policy.creatorForNetloc(host_used.encode('idna'), port_used)
        verifyObject(IOpenSSLClientConnectionCreator, creator)
```

Reply:
@rodrigc Given the traceback, why is the caller that's passing `hostname` as a `str` right?

Reply:
@exarkun I have submitted a separate patch, so that the test always passes bytes to createNetForLoc() here: https://github.com/LeastAuthority/txkube/pull/173 .  I have removed the use of `ensure_unicode()` here and gone back to `hostname.decode("ascii")`.